### PR TITLE
#2154 - Compute the overapproximation of the convex hull array of zonotopes with a zonotope recursively

### LINKED ISSUE #2154: Compute the overapproximation of  the convex hull array of zonotopes with a zonotope recursively
Add `overapproximate(::ConvexHullArray{N, ZT}, Zonotope) where {N, ZT<:AbstractZonotope{N}}`.

### PR DESCRIPTION
**Pull request by SebastianGuadalupe**
Closes #2154.

### CODE REVIEW

File: src/Approximations/overapproximate.jl [hunk 1]
```diff
@@ -1703,3 +1703,37 @@ function overapproximate(r::Rectification{N, <:AbstractZonotope{N}}, ::Type{<:Zo
...
+- `CHA` -- convex hull array of zonotopic sets
+- `Zonotope` -- type for dispatch
+
+### Output
+
+A zonotope overapproximation of the convex hull array of zonotopic sets.
+
+### Algorithm
+
+This function iterative applies the convex hull of two zonotopes
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1723]:
```suggestion
This function iteratively applies the overapproximation algorithm for the convex hull of two zonotopes to the given array of zonotopes.
```

File: src/Approximations/overapproximate.jl [hunk 2]
```diff
@@ -1703,3 +1703,37 @@ function overapproximate(r::Rectification{N, <:AbstractZonotope{N}}, ::Type{<:Zo
...
+
+This function iteratively applies the overapproximation algorithm for the convex hull of two zonotopes to the given array of zonotopes.
+"""
+function overapproximate(CHA::ConvexHullArray{N, <:AbstractZonotope{N}}, ::Type{<:Zonotope}) where {N}
+    arr = array(CHA)
+    n = length(arr)
+    if n == 1
+        return arr[1]
+    elseif n == 2
+        return overapproximate(ConvexHull(arr[1], arr[2]), Zonotope)
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1731]:
```suggestion
```
This case is covered by the `else` case already.

File: src/Approximations/overapproximate.jl [hunk 3]
```diff
@@ -1703,3 +1703,37 @@ function overapproximate(r::Rectification{N, <:AbstractZonotope{N}}, ::Type{<:Zo
...
+- `CHA` -- convex hull array of zonotopic sets
+- `Zonotope` -- type for dispatch
+
+### Output
+
+A zonotope overapproximation of the convex hull array of zonotopic sets.
+
+### Algorithm
+
+This function iteratively applies the overapproximation algorithm for the convex hull of two zonotopes to the given array of zonotopes.
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1723]:
```suggestion
This function iteratively applies the overapproximation algorithm for the
convex hull of two zonotopes to the given array of zonotopes.
```

**Comment by mforets** [comment]
The build in v1.4 passes. The build in v1.0 fails with

```julia
LazySets.Approximations.overapproximation: Error During Test at /home/travis/build/JuliaReach/LazySets.jl/test/unit_overapproximate.jl:267
  Test threw exception
  Expression: Z3 ⊆ Y
  MethodError: no method matching qr(::SparseMatrixCSC{Float32,Int64}, ::Type{Val{true}}; tol=7.510185f-5)
  Closest candidates are:
    qr(::AbstractArray{T,2}, ::Any) where T at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/qr.jl:363 got unsupported keyword argument "tol"
    qr(::SparseMatrixCSC; tol) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/SuiteSparse/src/spqr.jl:201
    qr(::AbstractArray{T,2}) where T at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/qr.jl:369 got unsupported keyword argument "tol"
  Stacktrace:
   [1] #qr#4(::Float32, ::Function, ::SparseMatrixCSC{Float32,Int64}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/SuiteSparse/src/spqr.jl:201
   [2] qr(::SparseMatrixCSC{Float32,Int64}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/SuiteSparse/src/spqr.jl:201
   [3] rank(::SparseMatrixCSC{Float32,Int64}) at /home/travis/build/JuliaReach/LazySets.jl/src/Arrays/matrix_operations.jl:24
   [4] #constraints_list#56(::Bool, ::Function, ::Zonotope{Float32,Array{Float32,1},SparseMatrixCSC{Float32,Int64}}) at /home/travis/build/JuliaReach/LazySets.jl/src/Interfaces/AbstractZonotope.jl:495
   [5] constraints_list at /home/travis/build/JuliaReach/LazySets.jl/src/Interfaces/AbstractZonotope.jl:490 [inlined]
   [6] _issubset_constraints_list(::Zonotope{Float32,Array{Float32,1},Array{Float32,2}}, ::Zonotope{Float32,Array{Float32,1},SparseMatrixCSC{Float32,Int64}}, ::Bool) at /home/travis/build/JuliaReach/LazySets.jl/src/ConcreteOperations/issubset.jl:300
   [7] issubset at /home/travis/build/JuliaReach/LazySets.jl/src/ConcreteOperations/issubset.jl:291 [inlined]
   [8] issubset at /home/travis/build/JuliaReach/LazySets.jl/src/ConcreteOperations/issubset.jl:317 [inlined] (repeats 2 times)
   [9] eval_test(::Expr, ::Expr, ::LineNumberNode) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:229
   [10] top-level scope at /home/travis/build/JuliaReach/LazySets.jl/test/unit_overapproximate.jl:267
   [11] include at ./boot.jl:317 [inlined]
   [12] include_relative(::Module, ::String) at ./loading.jl:1044
   [13] include(::Module, ::String) at ./sysimg.jl:29
   [14] include(::String) at ./client.jl:392
   [15] macro expansion at /home/travis/build/JuliaReach/LazySets.jl/test/runtests.jl:135 [inlined]
   [16] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
   [17] macro expansion at /home/travis/build/JuliaReach/LazySets.jl/test/runtests.jl:135 [inlined]
   [18] macro expansion at ./util.jl:156 [inlined]
   [19] top-level scope at /home/travis/build/JuliaReach/LazySets.jl/test/runtests.jl:135 [inlined]
   [20] top-level scope at ./none:0
```

**Comment by schillic** [comment]
The problem is with `Float32` and `Rational{Int}`. Simply move the tests to the loop `for N in [Float64]` at the end of the file.

**Pull request merged**